### PR TITLE
Support multi-tenant FHIR servers

### PR DESCRIPTION
**Pull request by reinkrul**
Requires an extra flag for HAPI FHIR server to enable partitioning, see README.md. Make sure to remove the volume of the "old" HAPI FHIR server Docker container, because otherwise partitions won't work.

Based on #67, must be merged first.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -141,18 +141,25 @@ func registerEHR(server *echo.Echo, config Config) {
...
+		return fhir.InitializeTenant(config.FHIR.Server.Type, config.FHIR.Server.Address, tenant)
+	}
 
 	if config.LoadTestPatients {
 		allCustomers, err := customerRepository.All()
 		if err != nil {
 			log.Fatal(err)
 		}
 		for _, customer := range allCustomers {
+			if err := tenantInitializer(customer.Id); err != nil {
```
Comment (woutslakhorst) [inline main.go:161]: The demo is currently able to use new customers without a restart, this would require a restart.

**Review by woutslakhorst — COMMENTED** [review]
Waiting for proxy PR adjustments.

File: domain/fhir/environment.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		return nil
+	}
+	if response.IsError() && response.StatusCode() != http.StatusNotFound {
+		return fmt.Errorf("error while checking for HAPI FHIR Server tenant (status-code=%d,tenant=%s): %s", response.StatusCode(), tenant, string(response.Body()))
+	}
+
+	// Tenant doesn't exist (yet), create it
+	idAsInt, err := strconv.Atoi(tenant)
+	if err != nil {
+		return fmt.Errorf("tenant is not an integer (tenant=%s): %w", tenant, err)
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply domain/fhir/environment.go:35]: Nope: https://github.com/nuts-foundation/nuts-registry-admin-demo/issues/59

File: domain/fhir/client.go [hunk 3]
```diff
@@ -122,3 +140,19 @@ func resolveResourcePath(resource interface{}) (string, error) {
...
+		if parts[i] == "" {
+			continue
+		}
+		// Make sure parts are separated by exactly 1 slash
+		if result != "" && !strings.HasSuffix(result, "/") && !strings.HasPrefix(parts[i], "/") {
+			result += "/"
+		}
+		result += parts[i]
+	}
+	println(result)
```
Comment (woutslakhorst) [inline domain/fhir/client.go:156]: leftover prinltn

File: domain/fhir/client.go [hunk 4]
```diff
@@ -122,3 +140,19 @@ func resolveResourcePath(resource interface{}) (string, error) {
 	}
 	return resourceType + "/" + resourceID, nil
 }
+
+func buildRequestURI(parts ...string) string {
```
Comment (woutslakhorst) [inline domain/fhir/client.go:144]: you could use `Path.Join`?

File: hapi-server.http [hunk 5]
```diff
@@ -0,0 +1,16 @@
+POST http://localhost:8080/fhir/DEFAULT/$partition-management-create-partition
```
Comment (woutslakhorst) [inline hapi-server.http:1]: some comments on how to use this file for the uninitiated?

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
small stuff

File: domain/fhir/client.go (hunk 4 above)
Comment (reinkrul) [reply domain/fhir/client.go:144]: I assumed it uses the OS-specific path separator, which wouldn't be correct for windows (`\`), but apparentely it always uses forward-slash (`/`), so probably: yes.

File: hapi-server.http (hunk 5 above)
Comment (reinkrul) [reply hapi-server.http:1]: Leftover, will remove

**Pull request merged**